### PR TITLE
Deprecate passing an object to .set

### PR DESCRIPTION
**Pull request by obaidott**
This PR addresses the first part of issue #245 

- [x] Create / document assign on DefineMap
- [x] Create / document update on DefineMap
- [x] Create / document deepAssign on DefineMap
- [x] Create / document deepUpdate on DefineMap
and..
- [x] Create / document assign on DefineList
- [x] Create / document update on DefineList
- [x] Create / document deepAssign on DefineList
- [x] Create / document deepUpdate on DefineList

### CODE REVIEW

File: list/list.js [hunk 1]
```diff
@@ -273,6 +275,106 @@ var DefineList = Construct.extend("DefineList",
...
+		assign: function(prop) {
+			if (canReflect.isListLike(prop)) {
+				canReflect.assignList(this, prop);
+			} else {
+				canReflect.assignMap(this, prop);
+			}
+			return this;
+		},
+		/**
+		 * @function can-define/list/list.prototype.assign update
```
Comment (phillipskevin) [inline list/list.js]: Should be `can-define/list/list.prototype.update update`

File: list/list.js [hunk 2]
```diff
@@ -273,6 +275,106 @@ var DefineList = Construct.extend("DefineList",
...
+		update: function(prop) {
+			if (canReflect.isListLike(prop)) {
+				canReflect.updateList(this, prop);
+			} else {
+				canReflect.updateMap(this, prop);
+			}
+			return this;
+		},
+		/**
+		 * @function can-define/list/list.prototype.assign assignDeep
```
Comment (phillipskevin) [inline list/list.js]: Should be `can-define/list/list.prototype.assignDeep assignDeep`.

File: list/list.js [hunk 3]
```diff
@@ -273,6 +275,106 @@ var DefineList = Construct.extend("DefineList",
...
+		 *
+		 * Sets an item or property or items or properties on a list.
+		 *
+		 * @signature `list.assignDeep(newProps)`
+		 *
+		 * Updates the properties on the list with `newProps`.
+		 *
+		 * ```js
+		 * var list = new DefineList(["A","B"]);
+		 * list.assignDeep({count: 1000, skip: 2});
```
Comment (phillipskevin) [inline list/list.js]: Can we make this example or description more clearly explain the difference between this and `assign`? Right now the example is identical.

File: map/map-test.js [hunk 4]
```diff
@@ -836,3 +837,107 @@ QUnit.test("non-Object constructor", function() {
...
+	});
+
+	obj.update({
+		list: ['another'],
+		foo: {
+			bar: 'zed'
+		}
+	});
+
+	QUnit.equal(obj.list.length, 1, 'list length should be 0');
```
Comment (phillipskevin) [inline map/map-test.js]: `list length should be 1`.

File: map/map-test.js [hunk 5]
```diff
@@ -836,3 +837,107 @@ QUnit.test("non-Object constructor", function() {
...
+
+	QUnit.equal(obj.list.length, 3, 'list length should be 3');
+
+
+	obj.assignDeep({
+		list: ['something']
+	});
+
+	QUnit.equal(obj.name, 'Test Name', 'Name property is still intact');
+	QUnit.equal(obj.list[0], 'something', 'list length should be 0');
```
Comment (phillipskevin) [inline map/map-test.js]: The list length should still be 3 here, right? I think you should deepEqual the entire list to see that only the first item was spliced.

File: map/map-test.js [hunk 6]
```diff
@@ -836,3 +837,107 @@ QUnit.test("non-Object constructor", function() {
...
+
+	QUnit.equal(obj.list.length, 3, 'list length should be 3');
+
+
+	obj.updateDeep({
+		list: ['something']
+	});
+
+	QUnit.equal(obj.name, undefined, 'Name property has been reset');
+	QUnit.equal(obj.list[0], 'something', 'list length should be 0');
```
Comment (phillipskevin) [inline map/map-test.js]: This assertion message is wrong also.

File: map/map-test.js [hunk 7]
```diff
@@ -183,15 +184,15 @@ QUnit.test("get with dynamically added properties", function(){
 
 QUnit.test("set multiple props", function(){
 	var map = new DefineMap();
-	map.set({a: 0, b: 2});
+	map.assign({a: 0, b: 2});
 
 	QUnit.deepEqual(map.get(), {a: 0, b: 2}, "added props");
 
-	map.set({a: 2}, true);
+	map.update({a: 2}, true);
```
Comment (phillipskevin) [inline map/map-test.js]: should this `, true` be removed?

File: list/list-test.js [hunk 8]
```diff
@@ -1132,6 +1132,50 @@ QUnit.test("can-reflect getKeyDependencies", function() {
 
 });
 
+QUnit.test("assign property", function() {
+	var list = new DefineList(["A","B"]);
+	list.assign({count: 0, skip: 2});
```
Comment (justinbmeyer) [inline list/list-test.js]: did you test assign when passed an array?

File: list/list.js [hunk 9]
```diff
@@ -273,6 +280,112 @@ var DefineList = Construct.extend("DefineList",
...
+		 * Assigns the properties on the list with `newProps`. Properties not present in `newProps` will be left unchanged.
+		 *
+		 * ```js
+		 * var list = new DefineList(["A","B"]);
+		 * list.assign({count: 1000, skip: 2});
+		 * list.get("count") //-> 1000
+		 * ```
+		 */
+		assign: function(prop) {
+			if (canReflect.isListLike(prop)) {
```
Comment (justinbmeyer) [inline list/list.js]: does `canReflect.assign` do the right thing here w/o you needing to feature test?

File: map/map.js [hunk 10]
```diff
@@ -184,6 +199,166 @@ var DefineMap = Construct.extend("DefineMap",{
 		return this;
 	},
 	/**
+	 * @function can-define/map/map.prototype.assignDeep assignDeep
+	 * @parent can-define/map/map.prototype
+	 *
+	 * @description Sets multiple properties on a map instance or a property that wasn't predefined.
```
Comment (justinbmeyer) [inline map/map.js]: what is meant by `wasn't predefined`

Comment (obaidott) [inline map/map.js]: I am not entirely sure, I took it from the existing documentation on `map.set(propName, value)`

https://canjs.com/doc/can-define/map/map.prototype.set.html

File: list/list-test.js (hunk 8 above)
Comment (obaidott) [inline list/list-test.js]: I added a test to reflect it.

File: map/map.js (hunk 10 above)
Comment (bmomberger-bitovi) [reply map/map.js]: I am guessing that "wasn't predefined" here indicates that there was no getter-setter added for that property yet (there was no definition on the prototype and the property wasn't in the constructing properties).

File: list/list.js [hunk 11]
```diff
@@ -1445,7 +1453,7 @@ Object.defineProperty(DefineList.prototype, "length", {
 			return;
 		}
 
-		if (newVal === this._length) {
+		if (newVal == null || isNaN(+newVal) || newVal === this._length) {
```
Comment (phillipskevin) [inline list/list.js]: Some comments explaining these cases would be helpful.

File: list/list.js [hunk 12]
```diff
@@ -1535,8 +1543,11 @@ canReflect.assignSymbols(DefineList.prototype,{
 	},
 
 	"can.deleteKeyValue": function(prop) {
+		prop = isNaN(+prop) || (prop % 1) ? prop : +prop;
```
Comment (phillipskevin) [inline list/list.js]: Again, a comment here would help a lot.

File: list/list.js (hunk 11 above)
Comment (bmomberger-bitovi) [reply list/list.js]: Done

File: list/list.js (hunk 12 above)
Comment (bmomberger-bitovi) [reply list/list.js]: And done

**Pull request merged**